service/storage_service.cc

```diff
@@ -2575,9 +2584,105 @@ class topology_coordinator {
+            co_await coroutine::maybe_yield();
+
+            auto& t = _db.find_column_family(table_id);
+            auto& rs = t.get_effective_replication_map()->get_replication_strategy();
+            if (!rs.uses_tablets()) {
+                continue;
+            }
+            const auto* nts_ptr = dynamic_cast<const locator::network_topology_strategy*>(&rs);
+            if (!nts_ptr) {
+                on_internal_error(slogger, "Cannot convert replication_strategy that uses tablets into network_topology_strategy");
+            }
+
+            auto rf_for_this_dc = nts_ptr->get_replication_factor(dc);
+            if (rf_for_this_dc <= 0) {
+                continue;
+            }
+            // Takes into account the RF of each DC, so stats will still make sense when aggregated
+            // from multiple DCs. This allows tablet allocator to compute an average tablet size
+            // by dividing total size in bytes by tablet count.
+            table_stats.size_in_bytes /= rf_for_this_dc;
+            slogger.info("raft topology: Refreshed table load stats for DC {}, table={}, RF={}, size_in_bytes={}, split_ready_phase={}",
+                          dc, table_id, rf_for_this_dc, table_stats.size_in_bytes, table_stats.split_ready_phase);
+        }
+
+        stats += dc_stats;
+    }
+
+    co_return std::move(stats);
+}
+
+future<> topology_coordinator::start_tablet_load_stats_refresher() {
+    auto can_proceed = [this] { return !_async_gate.is_closed() && !_as.abort_requested(); };
+    while (can_proceed()) {
+        bool sleep = true;
+        try {
+            _tablet_load_stats = make_lw_shared<const locator::load_stats>(co_await refresh_tablet_load_stats());
+            _topo_sm.event.broadcast(); // wake up load balancer.
+        } catch (raft::request_aborted&) {
+            slogger.debug("raft topology: Tablet load stats refresher aborted");
+            sleep = false;
+        } catch (seastar::abort_requested_exception) {
+            slogger.debug("raft topology: Tablet load stats refresher aborted");
+            sleep = false;
+        } catch (...) {
+            slogger.warn("Found error while refreshing load stats for tablets: {}, retrying...", std::current_exception());
+        }
+        if (sleep && can_proceed()) {
+            try {
+                co_await seastar::sleep_abortable(std::chrono::seconds(tablet_load_stats_refresh_interval), _as);
+            } catch (...) {
+                slogger.debug("raft topology: Tablet load stats refresher: sleep failed: {}", std::current_exception());
+            }
+        }
+    }
+}
+
 future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard guard) {
     slogger.debug("raft topology: Evaluating tablet balance");
 
+    if (utils::get_local_injector().enter("tablet_load_stats_refresh_before_rebalancing")) {
+        _tablet_load_stats = make_lw_shared<const locator::load_stats>(co_await refresh_tablet_load_stats());
```

Comment:
Why make_lw_shared?

Reply:
lifetime issues. tablet_allocator::balance_tablets() can yield and stats refresher fiber could override the one taken by balancer.

Reply:
makes sense, please add a comment where _tablet_load_stats is defined.